Google.Api.Gax.Grpc/GrpcCoreAdapter.cs

```diff
@@ -0,0 +1,165 @@
+﻿/*
+ * Copyright 2020 Google LLC
+ * Use of this source code is governed by a BSD-style
+ * license that can be found in the LICENSE file or at
+ * https://developers.google.com/open-source/licenses/bsd
+ */
+
+using Grpc.Core;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Reflection;
+using System.Threading;
+
+namespace Google.Api.Gax.Grpc
+{
+    /// <summary>
+    /// Implementation of <see cref="GrpcAdapter"/> for Grpc.Core.
+    /// </summary>
+    public sealed class GrpcCoreAdapter : GrpcAdapter
+    {
+        private const string GrpcCoreAssemblyName = "Grpc.Core, Version=2.0.0.0, Culture=neutral, PublicKeyToken=d754f35622e28bad";
+
+        // Option names, copied from ChannelOptions
+        internal const string ServiceConfigDisableResolution = "grpc.service_config_disable_resolution";
+        internal const string KeepAliveTimeMs = "grpc.keepalive_time_ms";
+        internal const string KeepAliveTimeoutMs = "grpc.keepalive_timeout_ms";
+        internal const string MaxReceiveMessageLength = "grpc.max_receive_message_length";
+        internal const string MaxSendMessageLength = "grpc.max_send_message_length";
+        internal const string PrimaryUserAgentString = "grpc.primary_user_agent";
+
+        // Note: we create the channel factory lazily rather than as part of type initialization as it can
+        // throw if the dependency is missing. Type initialization failing can make all kinds of things messy.
+        private static Lazy<Func<string, ChannelCredentials, GrpcChannelOptions, ChannelBase>> channelFactory =
+            new Lazy<Func<string, ChannelCredentials, GrpcChannelOptions, ChannelBase>>(CreateFactory, LazyThreadSafetyMode.PublicationOnly);
+
+        /// <summary>
+        /// Prevent lazy type initialization.
+        /// </summary>
+        static GrpcCoreAdapter()
+        {
+        }
+
+        /// <summary>
+        /// Returns the singleton instance of this class.
+        /// </summary>
+        public static GrpcCoreAdapter Instance { get; } = new GrpcCoreAdapter();
+
+        private GrpcCoreAdapter()
+        {
+        }
+
+        /// <inheritdoc />
+        protected override ChannelBase CreateChannelImpl(string endpoint, ChannelCredentials credentials, GrpcChannelOptions options) =>
+            channelFactory.Value.Invoke(endpoint, credentials, options);
+
+        private static Func<string, ChannelCredentials, GrpcChannelOptions, ChannelBase> CreateFactory()
```

Comment:
tiny nit: CreateChannelFactory

And also, I think a comment here with something like "A function is created, instead of just having a factory method, to avoid the reflection bits/expression building every time."

Reply:
SGTM.